ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateReaderSpec.scala

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import akka.NotUsed
+import com.google.protobuf.ByteString
+import akka.stream.scaladsl.{Sink, Source}
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
+  DamlLogEntry,
+  DamlLogEntryId,
+  DamlPartyAllocationEntry
+}
+import com.daml.ledger.participant.state.kvutils.Envelope
+import com.daml.ledger.participant.state.v1.{Offset, Update}
+import com.digitalasset.ledger.api.testing.utils.AkkaBeforeAndAfterAll
+import org.mockito.ArgumentMatchers._
+import org.mockito.Mockito.when
+import org.scalatest.mockito.MockitoSugar
+import org.scalatest.{AsyncWordSpec, Matchers}
+
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+
+class KeyValueParticipantStateReaderSpec
+    extends AsyncWordSpec
+    with Matchers
+    with MockitoSugar
+    with AkkaBeforeAndAfterAll {
+  "participant state reader" should {
+    "remove index suffix when streaming from underlying reader" in {
+      val reader = readerStreamingFrom(
+        offset = Some(Offset(Array(1, 1))),
+        LedgerRecord(Offset(Array(1, 2)), aLogEntryId, aWrappedLogEntry))
+      val instance = new KeyValueParticipantStateReader(reader)
+      val stream = instance.stateUpdates(Some(Offset(Array(1, 1, 0))))
+      val result = offsetsFrom(stream)
+
+      result.size shouldBe 1
+      result.head shouldBe Offset(Array(1, 2, 0))
+    }
+
+    "append index to internal offset" in {
+      val reader = readerStreamingFrom(
+        offset = None,
+        LedgerRecord(Offset(Array(1)), aLogEntryId, aWrappedLogEntry),
+        LedgerRecord(Offset(Array(2)), aLogEntryId, aWrappedLogEntry)
+      )
+      val instance = new KeyValueParticipantStateReader(reader)
+      val stream = instance.stateUpdates(None)
+      val result = offsetsFrom(stream)
+
+      result.size shouldBe 2
```

Comment:
Probably necessary; the next assertion will catch this and deliver a better error message.